blackbirds/mcmc.py

```diff
@@ -0,0 +1,223 @@
+        self.diff_mode = diff_mode
+        self.jacobian_chunk_size = jacobian_chunk_size
+        self.gradient_horizon = gradient_horizon
+        self.device = device
+        self.discretisation_method = discretisation_method
+        self._dim = self._verify_dim()
+        self._eye = torch.eye(self._dim)
+        self._previous_log_density = None
+        self._previous_grad_theta_of_log_density = None
+        self._proposal = None
+
+    def _verify_dim(self):
+        """
+        Checks the parameter dimension.
+        """
+        return self.prior.sample((1,)).shape[-1]
+
+
+    def _compute_log_density_and_grad(self,
+        data,
+        state
+    ):
+
+        _state = state.clone()
+        _state.requires_grad = True
+        ell = self.forecast_loss(data, _state)
+        log_prior_pdf = self.prior.log_prob(_state)
+        log_density = - ell + log_prior_pdf * self.w
+        grad_theta_of_log_density = autograd.grad(log_density, _state)
+        # Clip gradients if desired
+        if self.gradient_clipping_norm < np.inf:
+            total_norm = torch.linalg.vector_norm(grad_theta_of_log_density[0])
+            if total_norm > self.gradient_clipping_norm:
+                grad_theta_of_log_density = (grad_theta_of_log_density[0] * self.gradient_clipping_norm / total_norm,)
+                assert torch.isclose(torch.linalg.vector_norm(grad_theta_of_log_density[0]),
+                                     torch.Tensor([self.gradient_clipping_norm]))
+        return log_density.detach(), grad_theta_of_log_density
+
+    def initialise_chain(self,
+        data,
+        state
+    ):
+
+        _ = self._compute_log_density_and_grad(data, state)
+        self._previous_log_density = _[0]
+        self._previous_grad_theta_of_log_density = _[1]
+        self._proposal = None
+        return self._previous_log_density, self._previous_grad_theta_of_log_density
+
+    def step(self,
+        data,
+        current_state, 
+        scale: float = 1., 
+        covariance: torch.Tensor | None = None,
+        verbose: bool = False,
+    ):
+
+        """
+        Returns a (torch.Tensor, bool) pair corresponding to (the current state of the chain, whether
+        the current state resulted from an accept or reject decision in the Metropolis step).
+        """
+
+        if covariance is None:
+            covariance = self._eye
+        sC = scale * covariance
+        if self._previous_log_density is None:
+            # This would happen if the user hasn't initialised the chain themselves
+            out = self.initialise_chain(data, current_state)
+        if self.discretisation_method == 'e-m':
+            if self._proposal is None:
+                # This would happen if the user hasn't initialised the chain themselves
+                gradient_term = torch.matmul(sC, self._previous_grad_theta_of_log_density[0])
+                mean = current_state + gradient_term
+                if verbose:
```

Comment:
rather than having `verbose` and `print`, use the logging system.

```python
logger.debug(f"Total mean = f{mean}")
logger.debug(f"Gradient term = f{gradient_term}")

```